src/rdkafka_broker.c

```diff
@@ -2505,40 +2505,84 @@ static int rd_kafka_broker_ops_serve (rd_kafka_broker_t *rkb, int timeout_ms) {
 static void rd_kafka_broker_serve (rd_kafka_broker_t *rkb,
                                    rd_ts_t abs_timeout) {
         rd_ts_t now;
-        int initial_state = rkb->rkb_state;
         int remains_ms = rd_timeout_remains(abs_timeout);
+        int ops_timeout = RD_POLL_NOWAIT;
+        int io_timeout;
 
-        /* Serve broker ops */
-        if (rd_kafka_broker_ops_serve(rkb,
-                                      !rkb->rkb_transport ?
-                                      remains_ms : RD_POLL_NOWAIT))
-                remains_ms = RD_POLL_NOWAIT;
+        /* As a defensive measure, don't block forever.
+         * When blocking on the queue we will want to detect IO disconnects
+         * eventually.
+         * 1 wakeup/second is reasonable. */
+        if (unlikely(remains_ms == RD_POLL_INFINITE))
```

Comment:
Does it make sense to defend against all 'excessive' values for `abs_timeout` as opposed to just `RD_POLL_INFINITE`?  A value of 300000ms still seems fairly intrusive given we would wakeup infinite blocking calls every 1 second.

Reply:
Right, so a MIN(remains_ms, 1000) makes sense!